src/resources/ble_esp32.cc

```diff
@@ -1698,48 +1698,129 @@ PRIMITIVE(advertise_start) {
+        advertisement_size += additional_size;
+      }
+      const_cast<ble_uuid16_t*>(target_fields->uuids16)[target_fields->num_uuids16++] = uuid.u16;
     } else if (uuid.u.type == BLE_UUID_TYPE_32) {
-      uuids_32[fields.num_uuids32++] = uuid.u32;
+      // Make sure the additional UUID fits into the packet.
+      // For the first UUID we also have to include the 2 byte header of the list.
+      int additional_size = fields.num_uuids32 == 0 ? 6 : 4;
+      ble_hs_adv_fields* target_fields = &fields;
+      if (advertisement_size + additional_size > BLE_HS_ADV_MAX_SZ) {
+        fields.uuids32_is_complete = 0;
+        target_fields = &response_fields;
+        uses_scan_response = true;
+      } {
+        advertisement_size += additional_size;
+      }
+      const_cast<ble_uuid32_t*>(target_fields->uuids32)[target_fields->num_uuids32++] = uuid.u32;
     } else {
-      uuids_128[fields.num_uuids128++] = uuid.u128;
+      // Make sure the additional UUID fits into the packet.
+      // For the first UUID we also have to include the 2 byte header of the list.
+      int additional_size = fields.num_uuids128 == 0 ? 18 : 16;
+      ble_hs_adv_fields* target_fields = &fields;
+      if (advertisement_size + additional_size > BLE_HS_ADV_MAX_SZ) {
+        fields.uuids128_is_complete = 0;
+        target_fields = &response_fields;
+        uses_scan_response = true;
+      } {
+        advertisement_size += additional_size;
+      }
+       const_cast<ble_uuid128_t*>(target_fields->uuids128)[target_fields->num_uuids128++] = uuid.u128;
     }
   }
-  fields.uuids16 = uuids_16;
-  fields.uuids16_is_complete = 1;
-  fields.uuids32 = uuids_32;
-  fields.uuids32_is_complete = 1;
-  fields.uuids128 = uuids_128;
-  fields.uuids128_is_complete = 1;
+
+  if (name.length() > 0) {
+    int additional_size = 2 + name.length();
+    ble_hs_adv_fields* target_fields = &fields;
+    if (advertisement_size + additional_size > BLE_HS_ADV_MAX_SZ) {
+      // Without any name, there is no need to change the 'name_is_complete' field.
+      // We could cut the name and send a part of it, but that's not necessary.
+      fields.name = null;
+      target_fields = &response_fields;
+      uses_scan_response = true;
+    } else {
+      advertisement_size += additional_size;
+    }
+    target_fields->name = name.address();
+    target_fields->name_len = name.length();
+    target_fields->name_is_complete = 1;
+  }
 
   if (manufacturing_data.length() > 0) {
-    fields.mfg_data = manufacturing_data.address();
-    fields.mfg_data_len = manufacturing_data.length();
+    int additional_size = 2 + manufacturing_data.length();
+    ble_hs_adv_fields* target_fields = &fields;
+    if (advertisement_size + additional_size > BLE_HS_ADV_MAX_SZ) {
+      // Doesn't fit into the packet anymore.
+      // Keep it for the scan response.
+      target_fields = &response_fields;
+      // Without any mfg_data, there is no need to change the 'name_is_complete' field.
```

Comment:
This comment seems confusing. This is mfg_data, not name, and mfg_data dos not have a "complete"

Reply:
Removed. Bad copy/paste...